src/key_gen/mod.rs

```diff
@@ -0,0 +1,930 @@
+        // The row is valid. Encrypt one value for each node and broadcast a `Commitment`.
+        let mut values = Vec::new();
+        for (index, pk) in self.pub_keys.iter().enumerate() {
+            let val = row.evaluate(index + 1);
+            let ser_val = serialize(&FieldWrap(val))?;
+            values.push(sec_key.encrypt(pk, ser_val).ok_or(Error::Encryption)?);
+        }
+
+        let mut result = Vec::new();
+        result.push(Message::Commitment {
+            key_gen_id: self.our_index,
+            commitment: Commitment(sender_index, values),
+        });
+
+        if self.parts.len() == self.pub_keys.len() {
+            self.phase = Phase::Commitment;
+        }
+
+        Ok(result)
+    }
+
+    // Handles a `Commitment` message.
+    fn handle_commitment(
+        &mut self,
+        sec_key: &S,
+        sender_index: u64,
+        commitment: Commitment,
+    ) -> Result<CommitmentOutcome, Error> {
+        if self.phase != Phase::Commitment {
+            return Err(Error::Unknown);
+        }
+
+        let sender_id = self
+            .node_id_from_index(sender_index)
+            .ok_or(Error::UnknownSender)?;
+
+        Ok(
+            match self.handle_commitment_or_fault(sec_key, &sender_id, sender_index, commitment) {
+                Ok(()) => {
+                    if self.is_ready() {
+                        self.phase = Phase::Finalization;
+                    }
+                    CommitmentOutcome::Valid
+                }
+                Err(fault) => CommitmentOutcome::Invalid(fault),
+            },
+        )
+    }
+
+    /// Returns the index of the node, or `None` if it is unknown.
+    fn node_index(&self, node_id: &S::PublicId) -> Option<u64> {
+        self.pub_keys
+            .iter()
+            .position(|id| id == node_id)
+            .map(|index| index as u64)
+    }
+
+    /// Returns the id of the index, or `None` if it is unknown.
+    fn node_id_from_index(&self, node_index: u64) -> Option<S::PublicId> {
+        for (i, pk) in self.pub_keys.iter().enumerate() {
+            if i == node_index as usize {
+                return Some(pk.clone());
+            }
+        }
+        None
+    }
+
+    /// Returns the number of complete parts. If this is at least `threshold + 1`, the keys can
+    /// be generated, but it is possible to wait for more to increase security.
+    fn complete_parts_count(&self) -> usize {
+        self.parts
+            .values()
+            .filter(|part| part.is_complete(self.threshold))
+            .count()
+    }
+
+    /// Returns `true` if enough parts are complete to safely generate the new key.
+    fn is_ready(&self) -> bool {
+        self.complete_parts_count() >= self.threshold
+    }
+
+    /// Returns the new secret key share and the public key set.
+    pub fn generate_keys(&self) -> Result<(BTreeSet<S::PublicId>, Outcome), Error> {
+        if self.phase != Phase::Finalization {
+            return Err(Error::Unknown);
```

Comment:
Should we maybe have more specific error for this? Something like `OutcomeNotReady`. Or alternatively, change the return type to `Option` as there seems to be only two possibilities: either the outcome is available or it is not.

Reply:
right, will consider whether it will be a more explicit error or an option to be more suitable, within the next PR